### ISSUE TITLE
Array access not possible if number is to big when using ask

### ISSUE THREAD

**Issue by datWav** [piotrmurach/tty-prompt#83]
Hello there,

Awesome lib but sadly I run into a really nasty bug.
I asking after a UUID:
``` ruby
prompt = TTY::Prompt.new
uuid = prompt.ask('Request ID?')
```
With the input `14c3b412-e0c5-4ff5-9cd8-25ec3f18c702` (copy/paste or typing it in doesn't matter) the program is crashing.
```
Request ID? 14c3b412-e0c5-4ff5-9cd8-25ec3f18c7Traceback (most recent call last):
  27: from app.rb:7:in `<main>'
  26: from $HOME/coding/App/app/base.rb:30:in `run'
  25: from $HOME/coding/App/app/base.rb:30:in `loop'
  24: from $HOME/coding/App/app/base.rb:38:in `block in run'
  23: from $HOME/coding/App/app/base.rb:48:in `analyze_log'
  22: from $HOME/coding/App/app/base.rb:48:in `catch'
  21: from $HOME/coding/App/app/base.rb:65:in `block in analyze_log'
  20: from $HOME/coding/App/app/applications/gateway.rb:31:in `block in commands'
  19: from $HOME/coding/App/app/applications/gateway.rb:37:in `cmd_request'
  18: from $HOME/.rbenv/versions/2.5.0/gemsets/app/gems/tty-prompt-0.16.0/lib/tty/prompt.rb:175:in `ask'
  17: from $HOME/.rbenv/versions/2.5.0/gemsets/app/gems/tty-prompt-0.16.0/lib/tty/prompt.rb:154:in `invoke_question'
  16: from $HOME/.rbenv/versions/2.5.0/gemsets/app/gems/tty-prompt-0.16.0/lib/tty/prompt/question.rb:102:in `call'
  15: from $HOME/.rbenv/versions/2.5.0/gemsets/app/gems/tty-prompt-0.16.0/lib/tty/prompt/question.rb:111:in `render'
  14: from $HOME/.rbenv/versions/2.5.0/gemsets/app/gems/tty-prompt-0.16.0/lib/tty/prompt/question.rb:149:in `process_input'
  13: from $HOME/.rbenv/versions/2.5.0/gemsets/app/gems/tty-prompt-0.16.0/lib/tty/prompt/question.rb:160:in `read_input'
  12: from $HOME/.rbenv/versions/2.5.0/gemsets/app/gems/tty-reader-0.2.0/lib/tty/reader.rb:195:in `read_line'
  11: from $HOME/.rbenv/versions/2.5.0/gemsets/app/gems/tty-reader-0.2.0/lib/tty/reader.rb:390:in `trigger_key_event'
  10: from $HOME/.rbenv/versions/2.5.0/gemsets/app/gems/tty-reader-0.2.0/lib/tty/reader.rb:335:in `trigger'
   9: from $HOME/.rbenv/versions/2.5.0/gemsets/app/gems/wisper-2.0.0/lib/wisper/publisher.rb:42:in `broadcast'
   8: from $HOME/.rbenv/versions/2.5.0/lib/ruby/2.5.0/set.rb:338:in `each'
   7: from $HOME/.rbenv/versions/2.5.0/lib/ruby/2.5.0/set.rb:338:in `each_key'
   6: from $HOME/.rbenv/versions/2.5.0/gemsets/app/gems/wisper-2.0.0/lib/wisper/publisher.rb:43:in `block in broadcast'
   5: from $HOME/.rbenv/versions/2.5.0/gemsets/app/gems/wisper-2.0.0/lib/wisper/registration/object.rb:18:in `broadcast'
   4: from $HOME/.rbenv/versions/2.5.0/gemsets/app/gems/wisper-2.0.0/lib/wisper/broadcasters/send_broadcaster.rb:5:in `broadcast'
   3: from $HOME/.rbenv/versions/2.5.0/gemsets/app/gems/wisper-2.0.0/lib/wisper/broadcasters/send_broadcaster.rb:5:in `public_send'
   2: from $HOME/.rbenv/versions/2.5.0/gemsets/app/gems/tty-prompt-0.16.0/lib/tty/prompt/enum_list.rb:135:in `keypress'
   1: from $HOME/.rbenv/versions/2.5.0/gemsets/app/gems/tty-prompt-0.16.0/lib/tty/prompt/enum_list.rb:179:in `mark_choice_as_active'
$HOME/.rbenv/versions/2.5.0/gemsets/app/gems/tty-prompt-0.16.0/lib/tty/prompt/choices.rb:79:in `[]': bignum too big to convert into `long' (RangeError)
```

Reason for that is the line https://github.com/piotrmurach/tty-prompt/blob/0c24e622feab7d5a57e1e8f3362d1bfbf834220d/lib/tty/prompt/enum_list.rb#L134

As far as I unterstand it this function only needed if you want to mark a choices as active for example `enum_select`. But Why is it used for `ask`?

Best regards

**Comment by datWav** [comment]
Hi Piotr,

sorry I needed to travel and couldn't get back until now.

You are right. The code snippet worked. I didn't tested it back and thought.

But here is a new one.
```ruby
require 'tty/prompt'

prompt = TTY::Prompt.new
prompt.enum_select('Select something', ('A'..'Z').to_a)
uuid = prompt.ask('Request ID?')
```

For the `enum_select` you can choise a random number. After then try the same UUID for the ask.

![untitled](https://user-images.githubusercontent.com/924900/38817996-e6bb54a8-4199-11e8-8e0e-a67bd7081451.png)

Kontena looks interessting. I will have a look into it. ;)

**Comment by piotrmurach** [comment]
I've tried your 'new example' on Ruby 2.5.0 on Debian and MacOS and cannot replicate this, regardless of whenther I paste in the ID or type it out:
![tty-prompt-enum-list-with-ask-bug](https://user-images.githubusercontent.com/444312/38836784-585de440-41c7-11e8-9fd2-b2d77fb65ab9.png)
My only suggestion is to try installing Ruby again? Or you're doing something that is more than you present in your snippet. I really don't know what you're doing to cause this error!

**Comment by datWav** [comment]
I think you are missing some characters. It is only crashing when you type a 0 after the last 7. The initial UUID is a little longer then your input or try this input for `ask`: `123456789012345678901234567890`

I can reproduce it with the new snippet with:
64 bit - ArchLinux / Ubuntu 16.04
ruby 2.5.1 / 2.5.0
tty-prompt 0.16.0 and master

By the way, thank you for your help. ;)

**Comment by piotrmurach** [comment]
Initially, all the errors were pointing to `too large range` for the `enum_select`. Then, we have moved to `ask` being broken, which I cannot reproduce with any examples you have given me.

Even the ask input cited above for `123456789012345678901234567890` on Ruby `2.5.0` doesn't produce any errors whether pasted or typed manually, see for yourself 

![mystery](https://user-images.githubusercontent.com/444312/39083620-f73133f2-455e-11e8-891e-b7694bd41aa8.png)

There is really nothing to go on here. It points to either, the Ruby build being broken or you having some issue in your application that causes an error and the simple cases your showing do not reproduce the issue. I'm sorry I'm really busy and there is nothing for me to go on here or to fix. Maybe you can use https://asciinema.org/ to record a simple session where you input something manually and it reproduces the error?

**Comment by piotrmurach** [comment]
Hurray! I've managed to reproduce this bug and also found the reason! Thanks for persisting!

The number too big is a red hearing, and it is not actually what the bug is!

After digging around I've found a 'horrifying' 😱 thing! Every prompt such as `ask` and `enum_select` are short lived objects. However, they are never reclaimed by gc after prompt initialization because they subscribe to common events source provided by `tty-reader` which handles keyboard input. This is the reason why once a given keypress event is fired, all the prompts that still listen respond to any key events. This is obviously wrong, as once `enum_select` has been run it shouldn't listen for any events.

One way to solve this issue, would be to instationate reader for each prompt, or alternatively allow each prompt to unsubcribe once finished.
